packages/block-library/src/paragraph/test/paragraph.perf-native.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { fireEvent, measurePerformance, screen } from 'test/helpers';
+
+/**
+ * WordPress dependencies
+ */
+/**
+ * Internal dependencies
+ */
+import Paragraph from '../edit';
+
+describe( 'RichText Performance', () => {
+	it( 'performance is stable when typing text using the Paragraph block', async () => {
+		const scenario = async () => {
+			const paragraph = screen.getByPlaceholderText( 'Start writing…' );
+
+			fireEvent( paragraph, 'focus' );
+			fireEvent.changeText(
+				paragraph,
+				'<strong>Bold</strong> <em>italic</em> <s>strikethrough</s> text'
+			);
+		};
+
+		await measurePerformance(
+			<Paragraph
+				attributes={ {} }
+				setAttributes={ jest.fn() }
```

Comment:
The initial intent of this test was to create a deeper editor tree to create a more realistic editing experience using Paragraph block as an ancestor of RichText. The mocked props came from the Paragraph snapshot test, but you make a good point that they're not really testing a major cause of the performance problem while mocked. Given that the performance tests are not using the `initializeEditor` format like our other integration tests, do you see any obvious ways we could not mock these events in this test in order to test `setAttributes`? Perhaps something like this (in pseudocode), where [`RichText.onChange`](https://github.com/WordPress/gutenberg/blob/31885b5956c51be262062c39d86c9aadff1e2d3d/packages/rich-text/src/component/index.native.js#L1210) could fire [`RichText.onTextUpdate`](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/packages/rich-text/src/component/index.native.js#L1210) and then [`setAttributes`](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/packages/block-library/src/paragraph/edit.native.js#L71-L73)?

```
<Paragraph
	attributes={ { content } }
	setAttributes={ jest.fn() }
	onChange={ ( content ) => setAttributes( { content } ) }
/>
```

Alternatively, should we look at moving this test to a parent component that implements Paragraph as a child, or adding those as parent components here?

Reply:
The intent makes complete sense and aligns with our previous conversations. Reviewing the implementation merely revealed to me [the consideration](https://github.com/WordPress/gutenberg/pull/49221#discussion_r1147743705) I shared regarding uncovered portions of the performance issues. 

> Perhaps something like this (in pseudocode), where [RichText.onChange](https://github.com/WordPress/gutenberg/blob/31885b5956c51be262062c39d86c9aadff1e2d3d/packages/rich-text/src/component/index.native.js#L1210) could fire [RichText.onTextUpdate](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/packages/rich-text/src/component/index.native.js#L1210) and then [setAttributes](https://github.com/WordPress/gutenberg/blob/55e16d3c8a0e3b2ee8fd7f87a12cae8a2b1886d0/packages/block-library/src/paragraph/edit.native.js#L71-L73)?

If we mock `setAttributes` we could manually assert the number of times it is invoked. This could be an approach to track the number of time the store is updated or indirectly track subsequent renders that occur when attributes change. However, it would not provide a way to understand or measure the duration for renders that occur when attributes change. To test render durations, our test subject would need to be an ancestor — as you reference — similar to how `initializeEditor` renders a far deeper tree. 

It is is difficult to say which approach is valuable without knowing the optimization we would like to track. The current test structure could provide value if we want to track optimizations within `Paragraph` that are _not_ triggered by prop updates from ancestors. 

My hunch is that we will need to have our test subject be a deeper tree to monitor the impact of `onSelectionChange` and `setAttributes`. That would involve creating a form of [`initializeEditor`](https://github.com/WordPress/gutenberg/blob/169251a66542b99c2b0f76bca3be105dc04f3618/test/native/integration-test-helpers/initialize-editor.js#L28) that is compatible with Reassure. Example pseudo code: 

```js
await measurePerformance( <InitializeEditor initialHtml={ paragraphMarkup } />, { scenario: typeWithinParagraph } );
```

Reply:
> My hunch is that we will need to have our test subject be a deeper tree to monitor the impact of `onSelectionChange` and `setAttributes`. That would involve creating a form of [`initializeEditor`](https://github.com/WordPress/gutenberg/blob/169251a66542b99c2b0f76bca3be105dc04f3618/test/native/integration-test-helpers/initialize-editor.js#L28) that is compatible with Reassure.

I think this is a great idea, and would be a succinct way to generally test any block using Reassure if we could make it work. Given that Reassure is essentially a [thin render wrapper](https://github.com/callstack/reassure/blob/e564320cc93310206e8477e688b98fc2f40a3aff/packages/reassure-measure/src/measure.tsx#L60-L64) around React [`<Profiler />`](https://react.dev/reference/react/Profiler), I think we could modify the [equivalent render code](https://github.com/WordPress/gutenberg/blob/169251a66542b99c2b0f76bca3be105dc04f3618/test/native/integration-test-helpers/initialize-editor.js#L39-L44) in a new perf-test specific copy of initializeEditor to be compatible with Reassure's render implementation. I'll look at the output of [createElement and cloneElement](https://github.com/WordPress/gutenberg/blob/169251a66542b99c2b0f76bca3be105dc04f3618/test/native/integration-test-helpers/initialize-editor.js#L10) to see what else we'd need to modify to pass child elements to Reassure's render implementation.